### DISCUSSION TITLE
Enhancing the Layout UX of the Products block

### DISCUSSION THREAD

**Discussion by sunyatasattva** [woocommerce/woocommerce#44974] [WooCommerce, Blocks, and Site Editing]
Last week, as part of the [current phase of the Products block enhancements](https://github.com/woocommerce/woocommerce-blocks/issues/6793) being focused on UX improvements and polish, I took the challenge of implementing and enhancing new layouting options as described by @vivialice's design mock-ups. I thought that, compared to the other [large issue in this phase](https://github.com/woocommerce/woocommerce-blocks/issues/8135), this would be a lower-hanging fruit, but little did I know of the challenges I was going to face ahead.

This post serves as a little summary of the technical issues I faced during the implementation, and also as a place to discuss the next steps and take feedback and decisions, as I did not end up implementing the feature exactly as described in @vivialice's specs, for reasons that I'll try to clarify below.

> **Note**
> I've opened a draft PR containing the implementation I settled in with, in the hopes that trying things out could be a good way to clarify the context of this discussion.
>
> :octocat: **PR link:** https://github.com/woocommerce/woocommerce-blocks/pull/8274 

## The original idea and what stood in the way

<img align="left" width="214" alt="Screenshot 2023-01-24 at 00 04 42" src="https://user-images.githubusercontent.com/1847066/214172147-849e7e85-60be-4bfb-9bb9-29e9bccd0286.png">

As you can see on the mock-up on the left, the original idea was to enhance the pre-existent “Layout” section with options that would make sense to be there. The core Query Loop block, upon which the Products (Beta) block is based on, has only one controls in this section:

**Inner blocks use content width**: is an option that neither me nor @vivialice were able to understand fully. The mock-ups still keep it, but I guess this is a good time to question whether we want it for our block, or we want to remove it or, perhaps, rephrase it (if it's unclear to both me and Viv, I suspect this might be unclear to most users too).

The **Columns** control (which controls) the number of columns in case the Query Loop block is set up to be displayed in “grid-mode” is unintuitively bunched within the “Settings” (which are more content-related, such as what Post Type to display, and whether to include Sticky posts etc.).

Unfortunately, the number of rows is also not controlled from the same place: instead, a control of maximum number of products to show per page is present in the toolbar, something that has recently confused even members of our own team.

Viv's design aimed to fix that and centralize options affecting the layout in this section, and also add some extra options which are not afforded by the Query Loop (but which could perhaps be an improvement to the Query Loop itself, rather than our own variation, as I can see them being useful in general), plus support for more features such as padding, margin and color styles.

**So, what is the problem?**

### Extending the Inspector Controls

The problem is similar to one we have encountered at the beginning of our project and that we are still struggling with (for example, with @imanish003's work on #8267 and #8268): namely, the extensibility capabilities currently provided by Gutenberg, are not enough to implement our vision.

Last year, we have added the capability of [certain controls of the Query Loop block to be programmatically disabled by variations](https://github.com/WordPress/gutenberg/pull/43632), but that was an _ad hoc_ functionality and is implemented on a control-by-control basis. In this case, we have no way to extend, hook into, or remove and disable the existing “Layout” section within the Inspector Controls. Such extensibility [has been a topic in the Gutenberg community for a while](https://github.com/WordPress/gutenberg/issues/6023), and I believe we can do something to push the envelope and take the opportunity to contribute upstream and to the broader community.

A (dirty) solution I could have taken, could have been to remove the existing “Layout” section through some brute-force DOM manipulation method (or CSS—like it has been done before by other projects trying to extend Gutenberg), but, besides the obvious drawback of it being dirty, non-scalable, and potentially frail code, we also would have lost the existing control from GB core for the columns. I mean, it is a small thing to implement, but the entire reason we went with the variation idea was to use as much as possible what was offered to us, and make our block future proof. So, in my opinion that was not the way to go.

However, certain standard panels already support hooking into them for the purpose of extending things, which I think is great! Moreover, this seems to be a solution the Gutenberg team is open to supporting (for example, see [this experimental draft PR](https://github.com/WordPress/gutenberg/pull/43684) by @ntsekouras implementing something similar). The layout panel is also one of the “standard” panels added to the Inspector controls, and I believe that many blocks would benefit from leveraging it to add their own layouting options.

So, I have created a [draft PR in my own fork of Gutenberg](https://github.com/sunyatasattva/gutenberg/pull/5) to test this assumption and play around. The necessary changes for this are really just a few lines of code, and they are also absolutely non-destructive. I strongly believe this is something worth contributing upstream.

### Enabling styles controls

The design spec introduced styling controls of colors, padding and margin on the Products (Beta) block which is usually done by [adding some simple records to the `supports` prop of the `block.json`](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/). Unfortunately, the [Variations API](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-variations/) does not allow variations to edit or override the `supports` property.

I think it's pretty clear that we are using the Variation API in a way in which it wasn't exactly conceived to be used. However, I don't think this is a bad thing: by highlighting these use-cases and exploring what's possible, I strongly believe we can slowly expand the surface area of the extensibility as a whole.

To be honest, while my first gut reaction was that I wished the Variation API would provide the capability to extend the `supports`, after thinking about it for a while, I now think that perhaps this might not be great conceptually. It is probably responsibility of the parent block to provide and define such capabilities, and, as variation developers, if we have a case for that, we might want to contribute to the core block itself.

I don't actually have a strong opinion about this and am still struggling with this though, so any ideas on this would be appreciated.

Nevertheless, another interesting part I discovered while implementing this spec, was that style options for colors were present already. However, they were recently [purposefully removed](https://github.com/WordPress/gutenberg/pull/46147) and have been applied to the “Post template” inner child of the Query Loop block, rather than on the Query Loop itself. Reading the context of that PR and quoting from that:

> The Query block is to be a "settings only" block. Applying the colors at such a high level has inconsistent results as well.

I wonder whether this is a good decision, as I can see people being slightly confused on where to find certain styling options, as the “Post template” is a bit of a difficult block to wrap your head around if you are a non-technical user, I think.

In any case, implementing those would probably be against Gutenberg's design decision, so perhaps we shouldn't go ahead in any case with those until we have somehow bridged this conceptual gap.

### The “gap“ sliders

Another part of the design spec, were the “column” and “row spacing” sliders. First thing I noticed, is that the design spec designated them to have inputs on the left and the slider on the right side. However, the `RangeControl` component used in Gutenberg, [has the opposite design](https://wordpress.github.io/gutenberg/?path=/story/components-rangecontrol--default): slider on the left and input on the right.

However, Viv had obviously taken those components from somewhere and I realized that actually most other modern sliders in Gutenberg (e.g. the slider used in Margin and Padding controls) have the design that Viv had proposed (i.e. opposite of the typical `RangeControl` component) and are not strictly available as exported components from Core.

This creates a bunch of UX inconsistency, which would be most visible in the implementation of this spec because those different components would be one next to another and the relative positions of their inner components would be swapped.

I thought one idea could have been to allow the `RangeControl` component to have an option to decide where to put the input relative to the slider (such positional options are not uncommon in Core components, actually, so it would fit the architecture philosophy). However, after talking with the Gutenberg design team, I was told by @mirka that the `RangeControl` component is in the process of being deprecated:

> We're in the process of migrating RangeControl to a simpler Slider component that [doesn't have UnitControl built in](https://github.com/WordPress/gutenberg/pull/42315#issuecomment-1191752373). From there, we'll build some kind of higher-level component that composes Slider with UnitControl/NumberControl.

> So while we won't be adding new APIs to RangeControl anymore, we'll likely be able to add left/right support into the new composed components.

> In any case we're still months away from getting there, so if you're in a hurry your best bet might be to compose your own UnitControl + RangeControl component if you want it on the left side.

On top of that, during our design discussion with Viv, I proposed that perhaps it would actually make sense to use the same control that it is used for other spacing options (such as Margin and Padding) and provide also the “stepper” option with sensible presets. 

After she agreed to this change and I started researching the “stepper” control, I realized that the “Row Spacing” and “Column Spacing” would basically just be what Core calls “Block Spacing”, which is a global style option available in certain blocks which controls just that (though the wording is “Horizontal” and “Vertical” rather than “Row/Column”).

<img align="left" src="https://user-images.githubusercontent.com/1847066/214180732-9d53aa38-c7b4-4548-88c0-61c660d04ca0.png" />

This would be the same issue as described above, however: variations can't edit the block `supports` property. Also, even if I was able to edit that, those styles _technically_ belong to the “Post template” block once again, as this is the wrapper element that decides the gaps. As such, because of [Limitations of the Block Supports API](https://github.com/WordPress/gutenberg/blob/trunk/docs/explanations/architecture/styles.md#current-limitations-of-the-block-supports-api), it would not easily be possible to apply those styles.

In the interest of exploration, I tried some solutions and you can see that if you use the custom GB branch above, I actually enabled the block supports for this property and played around with it. Also, in order to have those sliders as they are in the spec (belonging to the Layout section) I reproduced the components and managed to make them work.

For the reason explained above, however, styles were not correctly applied in the editor, and I tried some really crazy workarounds to make it work, including using Global styles from `@emotion/css` to generate and apply them. I won't get into the details of the technical difficulties involved here, but I eventually settled for a solution which is also pretty out there and involves creating a `style` element and using portals to place it correctly (a solution that also has a precedent in Gutenberg, by the way). You can check it out in my PR linked at the beginning of this post.

## So, what's next?

The goal of this post was to outline the technical difficulties and my journey implementing those settings. During this journey, I discovered that there are some clashes between our vision, what's currently possible, and also some of the explicit ideas from the Gutenberg team.

So, I wished to kick off this conversation, to flatten out these differences. I believe these are actually belonging to three categories:

1. Things which are not currently supported by Gutenberg, but which likely align with Gutenberg vision, and could benefit the larger community. For this category, I believe the best solution is to just contribute upstream. One example of an item in such category would be the Layout panel extensibility.
2. Things which are not currently supported by Gutenberg, and clash with explicit decisions from the Gutenberg team. For this category, I propose that we try to follow the direction that Gutenberg took. If we disagree with a decision, I also believe that it could be worth it to make a case and discuss it out with them, to see whether we can champion our point of view. During my research, I noticed that obviously some decisions are made as well because the full context of our specific use-case is clearly not under their radar. Some folks at GB know what we are doing, but most don't, and it'd be beneficial to share our points of view with them. One such example, would be the styling options.
3. Things which are not considered by Gutenberg and are specific to us, or that are far away in the future on the Gutenberg timeline. In this case, we should probably implement those ourselves. Likely, the gap sliders are one of those examples.

As usual, I'm looking forward to feedback and opinions.

**Comment by nerrad** [comment]
Thanks for outlining the challenges you've encountered and some proposed outcomes! A few comments (I'm going to put as individual comments so they can be threaded):

Re "Inner blocks use content width setting", Nick Diego has a good explanation in [this article](https://wpengine.com/builders/exploring-block-layout-alignment-and-dimensions/#:~:text=When%20%E2%80%9CInner%20blocks%20use%20content,designs%20call%20for%20unique%20layouts.)

> So, I have created a https://github.com/sunyatasattva/gutenberg/pull/5 to test this assumption and play around. The necessary changes for this are really just a few lines of code, and they are also absolutely non-destructive. I strongly believe this is something worth contributing upstream.

I'm good with that, I'm assuming the plan would be to just keep the current UI/UX until something like that lands?

Specific to the draft PR, might be good to try splitting it up a bit - it looks like the design tools support could be separate from the range controls move? 

Since WP 6.2 code freeze is coming up soon (first week of February) I would publish your PR soon so at least there could be potential discussion around it (do a check on the existence of any issues/PRs for adding design tools support because I think that is an ongoing initiative).

**Reply by nerrad** [reply]
> I wonder whether this is a good decision, as I can see people being slightly confused on where to find certain styling options, as the “Post template” is a bit of a difficult block to wrap your head around if you are a non-technical user, I think.
> 
> In any case, implementing those would probably be against Gutenberg's design decision, so perhaps we shouldn't go ahead in any case with those until we have somehow bridged this conceptual gap.

So does this mean that you don't think the "supports" property you added for color and spacing should be added to the query block after-all? If so, I think I agree here, especially given they were intentionally removed. 

With that said, I also agree there is room for contributing some feedback to the GB project on your findings (maybe initially as a discussion/issue) and proposing alternatives if you have ideas in mind. For example, one thing I've often wondered/stumbled over myself is the overloaded use of the term "Template" (for the "Post Template" block). It can be confusing with how the word "template" is used elsewhere on the site and block editors. I wonder if it'd be better to change that to "Post Layout" (and automatically update the block name for different post types in use).

**Reply by sunyatasattva** [reply]
> Re "Inner blocks use content width setting", Nick Diego has a good explanation in [this article](https://wpengine.com/builders/exploring-block-layout-alignment-and-dimensions/#:~:text=When%20%E2%80%9CInner%20blocks%20use%20content,designs%20call%20for%20unique%20layouts.)

Thank you! That helps. I guess the example Nick makes about making the background of one full-width color and then making sure inner blocks maintain layout width would be another point in favor of the idea of having colors scoped to the Query Loop block, what do you think? Or, alternatively, to move this layouting option to the Post Template, where color styles currently live. 

> I'm good with that, I'm assuming the plan would be to just keep the current UI/UX until something like that lands?

I guess that's another thing I'm open to feedback for. We could:

1. Bruteforce the extension of the panel via some dirty DOM manipulation.
2. Have the options available in a different panel until something like that lands.

The advantage of (1.) is that transition will be invisible to the user. Drawback is that we'll have to get some frail and ugly code shipped. I'm open to suggestions.

> Specific to the draft PR, might be good to try splitting it up a bit - it looks like the design tools support could be separate from the range controls move?

Yep, 💯 . That's why I didn't even open that PR even as a draft in GB. It needs to be split up in the minimal code required for the proposed change. The reason why I left it as is, is to make testing of my other PR easier and keep code changes in one place to see more or less what is that I'm proposing and I have explored.

**Reply by sunyatasattva** [reply]
> So does this mean that you don't think the "supports" property you added for color and spacing should be added to the query block after-all? If so, I think I agree here, especially given they were intentionally removed.

I think there is a case to be made for that, yes. I'd probably trust @vivialice's opinion as the UX expert more than my own, and we'd need to talk with people who made that decision as there is likely blind spots we don't see now. Currently, I think that everybody in our team I spoke to (at the meetup and during projects) agrees that the “Post Template” is a bit of an unintuitive experience and should be used as little as possible (we even discussed the possibility of having “invisible” or “unselectable” blocks in the list view).

After reading those PRs, however, it seemed to me that Gutenberg is switching to an idea where the Query Loop is the wrapper controlling the settings of the query, while the Post Template is the wrapper controlling the style/aesthetics. 

> With that said, I also agree there is room for contributing some feedback to the GB project on your findings (maybe initially as a discussion/issue) and proposing alternatives if you have ideas in mind. For example, one thing I've often wondered/stumbled over myself is the overloaded use of the term "Template" (for the "Post Template" block). It can be confusing with how the word "template" is used elsewhere on the site and block editors. I wonder if it'd be better to change that to "Post Layout" (and automatically update the block name for different post types in use).

Agree with this sentiment.

**Comment by nerrad** [comment]
> So, I wished to kick off this conversation, to flatten out these differences. I believe these are actually belonging to three categories:

> Things which are not currently supported by Gutenberg, but which likely align with Gutenberg vision, and could benefit the larger community. For this category, I believe the best solution is to just contribute upstream. One example of an item in such category would be the Layout panel extensibility.
> Things which are not currently supported by Gutenberg, and clash with explicit decisions from the Gutenberg team. For this category, I propose that we try to follow the direction that Gutenberg took. If we disagree with a decision, I also believe that it could be worth it to make a case and discuss it out with them, to see whether we can champion our point of view. During my research, I noticed that obviously some decisions are made as well because the full context of our specific use-case is clearly not under their radar. Some folks at GB know what we are doing, but most don't, and it'd be beneficial to share our points of view with them. One such example, would be the styling options.
> Things which are not considered by Gutenberg and are specific to us, or that are far away in the future on the Gutenberg timeline. In this case, we should probably implement those ourselves. Likely, the gap sliders are one of those examples.
As usual, I'm looking forward to feedback and opinions.

I think that's a good categorization. 

On the whole, however, I think there are, some cases where we might not be fully aware of the design direction for GB and what is already planned/on the horizon. @vivialice have you synced up with the GB design team to validate the direction on some of the things Lucio brought up in this post?

**Comment by nerrad** [comment]
Just to make sure I'm understanding correctly Lucio, what are you looking for as an outcome of this discussion?

**Reply by sunyatasattva** [reply]
I apologize for the lack of clarity. I'm looking for two outcomes:

## Specifically to this discussion

Whether I'm going in the right direction and I should proceed with the proposals I made. That is:

1. Open a PR in Gutenberg to allow extending the Layout section of the Inspector Controls.
2. Whether we should add styles and dimension controls (as per @vivialice's specs), and whether we should:
   a. Propose adding support for them in the Query Loop or Post Template.
   b. Propose making it possible for the variation API to extend the `supports` prop of the parent block.
   c. I suppose we all agree that force-implementing our own would be a bad idea and we should discard that, right?
3. What should we do re: the block spacing:
   a. Proceed with the implementation of the “Block spacing” in my PR, even though the code is a bit convoluted.
   b. Bunch it up with point 2, above.

<small>Note that the reason why I consider 2c to be off the table, while 3a could be acceptable, is because adding block spacing would not be a simple matter of adding a `supports` line. In my experiment, extra code is still needed to make that work correctly, while for colors, and padding/margin it works out of the box.</small>

## Generically

As we are moving forward with the core philosophy of trying as much as possible to tightly integrate with the upstream work in Gutenberg, I feel we are, in general, not aligned enough between the projects. I understand Gutenberg is a larger project and has many more contributors, active conversations and even interest groups.

However, currently it seems that we are mostly playing catch-up, and sometimes even losing time specc'ing against explicit upstream decisions (such as in the style case).

I think that, internally, it'd be good if we aligned on the three categories above, and how to identify them and what to do as rules of thumb in each of the cases: as the process is currently a bit too much _ad hoc_, I think it would benefit us to have some agreement on how to behave, that could have guided—for example—both my question here, and @imanish003's question about proposing a “Buttons” extension to allow disabling editing of the content.

Externally, it would be good to have much more structured alignment opportunities with Gutenberg. I think what you said above to be spot on:

> On the whole, however, I think there are, some cases where we might not be fully aware of the design direction for GB and what is already planned/on the horizon. @vivialice have you synced up with the GB design team to validate the direction on some of the things Lucio brought up in this post?

It'd be perhaps useful to have one of us to be a sort of “Gutenberg” ambassador, whose role would be to keep up with the conversations in Gutenberg, discuss our needs, champion our use-cases and find the right people to do something. This person could perhaps, if permitted, join Gutenberg design meetings and such events. Alternatively (or, why not both?), someone from Gutenberg could meet regularly with us (perhaps fortnightly on our design meetings with Viv) to help validate ideas and make sure we don't follow some red herring.

What do you think?

**Comment by nerrad** [comment]
I think it'll be helpful to sum up my thoughts in a single comment rather than thread them. 

### Layout Panel PR

While I'm partial to the improvements I think these changes will make, before we submit any PR to move settings like row and columns to a new layout panel - there needs to be alignment with the GB design team around this. So I think @vivialice needs to connect with the relevant folks and get their feedback.

### Styles and Dimensions Controls (and Block Spacing)

Similar to the Layout issue, this needs some alignment with the GB design team. @vivialice, this is a good thing to raise with them as well.

---
Before we commit time to the technical implementations of the above things (which I agree include some done at the GB level), I think we need to make sure we clearly understand and have alignment with the general direction of the GB design team around controls and design tools that could feasibly be included in the WP Query Block. @vivialice has this been done yet?

Then, as a part of this, we'll have a better idea of those things that might be specific to the products block and where we'd need extensibility.


> It'd be perhaps useful to have one of us to be a sort of “Gutenberg” ambassador, whose role would be to keep up with the conversations in Gutenberg, discuss our needs, champion our use-cases and find the right people to do something. This person could perhaps, if permitted, join Gutenberg design meetings and such events. Alternatively (or, why not both?), someone from Gutenberg could meet regularly with us (perhaps fortnightly on our design meetings with Viv) to help validate ideas and make sure we don't follow some red herring.

Not disagreeing but I don't think that's a conversation for this thread.

**Reply by sunyatasattva** [reply]
Thanks for the reply.

I guess I'll put things on hold, then, until a decision is met. However,

> While I'm partial to the improvements I think these changes will make, before we submit any PR to move settings like row and columns to a new layout panel - there needs to be alignment with the GB design team around this. 

Don't you think this PR could be valuable in general regardless of this decision? It seems like that the Layout section is “the odd one out” as other globally added sections all have slots integrated.

We talked with Viv earlier today and we'll wait for her discoveries, then!

**Comment by danieldudzic** [comment]
> I think it's pretty clear that we are using the Variation API in a way in which it wasn't exactly conceived to be used.

I agree 100%. This became very apparent to me when looking into adding Global Styles support for Variations: https://github.com/WordPress/gutenberg/issues/47324 - it was a bit of a surprise finding this out, I was expecting the Variations API simply building on top of the current Blocks API (and all the features available in Blocks).

It turned out that rather than Variations being 'full-fledged' blocks inheriting functionality but having a lot of independence, they are currently viewed as block descendants closely tied to their parent, serving mostly the same purpose, just adding a different "flavor".

> However, I don't think this is a bad thing: by highlighting these use-cases and exploring what's possible, I strongly believe we can slowly expand the surface area of the extensibility as a whole.

I believe that the concept of (Block) Variations (in core) will need to evolve in order to accomodate Woo-specific functionality. Current limitations are a bit frustrating and are definitely slowing us down in terms of shipping planned features.

I'm wondering if the decision to start using variations (rather than stand-alone blocks) wasn't a bit premature given the amount of time that will still be needed to evolve the Variations API in core. This isn't a strong opinion of mine, just something popping into mind from time to time.
